### PR TITLE
[AAP-76053] fix: handle retries on transient outages

### PR DESCRIPTION
**Pull request by mkanoor**
## What

Add exponential-backoff retry logic across all external connections
(PostgreSQL, Controller API, WebSocket) so that ansible-rulebook
survives transient infrastructure outages instead of crashing.

## Why

During database failovers, controller restarts, or websocket drops,
ansible-rulebook would crash immediately on the first connection
error. This forces operators to manually restart the process after
every transient outage.

Jira: AAP-76053

## Changes

**pg_listener** — tenacity-based async retries with exponential backoff,
TCP keepalives enabled by default (~40s dead connection detection),
auth failure fast-path (no retry on bad credentials).

**job_template_runner** — aiohttp-retry RetryClient for GET
(429/500/502/503/504) and POST (429/502/503) with
ClientConnectorError, configurable via `--controller-retry-max-timeout`
and `--controller-retry-attempts` CLI args /
`EDA_CONTROLLER_RETRY_MAX_TIMEOUT` and `EDA_CONTROLLER_RETRY_ATTEMPTS`
env vars, availability state tracking to avoid log spam.

**websocket** — retry on all ConnectionClosedError codes including 1011.

**Dependencies** — aiohttp-retry >=2.9,<3 and tenacity >=9,<10.

**Other fixes**
- `close_session()` sets `self._session = None` to avoid reusing closed sessions
- `_launch` and `_create_obj` parse response body before `raise_for_status()` to preserve the "already exists" idempotent-create path
- `conftest.py` patches aiohttp 3.14 `stream_writer` compatibility (replaces forked aioresponses dependency)
- `requirements_test.txt` reverted to stock aioresponses
- `tests/e2e/utils.py` uses explicit `raise RuntimeError` instead of `print(data["bad"])`
- `connection_attempts` type hint fixed to `Optional[list]` (PEP 484)

## Test plan

- [x] pg_listener: retry-then-recover, auth failure fast-path,
      keepalive defaults/overrides, backoff reset between failovers
- [x] Controller: 429/502/503/504 retry/recovery, max-retry exhaustion,
      unavailable/restored logging transitions
- [x] WebSocket: 1011 retry e2e test, unit test updated
- [ ] CI pipeline passes

Assisted-by: Claude Code / Opus 4.6 (Anthropic)

## Summary by CodeRabbit

* **New Features**
  * CLI options to configure controller retry timeout and attempts
  * PostgreSQL listener supports reconnect/backoff and TCP keepalive configuration

* **Improvements**
  * Controller interactions and WebSocket flows use exponential backoff, retries, and clearer restored/unavailable logging
  * Job-controller handling tolerates transient/invalid JSON responses more gracefully
  * WebSocket retry logic refines which close codes are treated as transient

* **Documentation**
  * CLI help documents new retry options

* **Tests**
  * E2E and unit tests expanded for retry/reconnect scenarios

* **Chores**
  * Added runtime dependency for HTTP retry support

### CODE REVIEW

File: requirements_test.txt [hunk 1]
```diff
@@ -21,6 +21,7 @@ freezegun
 oauthlib>=3.2.0
 kubernetes
 urllib3 >= 2.6.3
-aioresponses
+# TODO: revert to 'aioresponses' after https://github.com/pnuckowski/aioresponses/issues/289 is released
+aioresponses @ git+https://github.com/agners/aioresponses.git@fix-aiohttp-3.14-stream-writer
```
[the remark this answers is not in the document]
Comment (mkanoor) [reply requirements_test.txt]: Fixed

File: tests/test_websocket.py [hunk 2]
```diff
@@ -283,9 +286,11 @@ async def test_send_event_log_to_websocket_with_non_recoverable_exception(
...
+    socket_mock.return_value.send.side_effect = [
+        websockets.exceptions.ConnectionClosedError(rcvd=rcvd, sent=None),
+        data_sent.append({"a": 1}),
+        data_sent.append({"b": 2}),
+    ]
 
-    with pytest.raises(websockets.exceptions.ConnectionClosedError):
-        await send_event_log_to_websocket(queue)
+    await send_event_log_to_websocket(queue)
+    assert data_sent == [{"a": 1}, {"b": 2}]
```
[the remark this answers is not in the document]
Comment (mkanoor) [reply tests/test_websocket.py]: Fixed

File: ansible_rulebook/job_template_runner.py [hunk 3]
```diff
@@ -120,8 +135,27 @@ async def _get_page(self, href_slug: str, params: dict) -> dict:
...
+            )
         except aiohttp.ClientError as e:
+            if self._controller_available:
+                self._controller_available = False
+                logger.warning(
+                    "Controller unavailable (%s), " "requests will be retried",
+                    e,
+                )
             logger.error(CLIENT_CONNECT_ERROR_STRING, str(e))
             raise ControllerApiException(str(e))
```
[the remark this answers is not in the document]
Comment (mkanoor) [reply ansible_rulebook/job_template_runner.py:185]: Fixed

**Comment by wfealdel** [comment]
@coderabbitai resume

**Comment by wfealdel** [comment]
@mkanoor 

### Critical Finding

**`_create_obj` "already exists" check broken by `raise_for_status()` reordering** — `job_template_runner.py`

The PR swapped the order of body parsing and status checking in both `_create_obj` and `_launch`. The original code parsed the response body *before* calling `raise_for_status()`:

```python
# Original (main branch) — _create_obj
body = json.loads(await post_response.text())   # body populated first
post_response.raise_for_status()                 # raises on 400, but body is set
```

The PR reverses this:

```python
# PR — _create_obj
post_response.raise_for_status()                 # raises on 400 immediately
response_text = await post_response.text()       # never reached on error
body = json.loads(response_text)                 # body stays None
```

When the controller returns HTTP 400 with `"already exists"` (concurrent label creation in HA), `raise_for_status()` throws before `body` is assigned. The exception handler evaluates `"already exists" in str(None)` — always `False`. This breaks the idempotent-create retry path in `_get_or_create_label`.

**Same pattern in `_launch`:** the `if body: logger.error("Error: %s", body)` block in the `ClientError` handler is now dead code.

**Suggested fix:** parse the body before `raise_for_status()`, preserving the original order:

```python
async with self._session.post(...) as post_response:
    response_text = await post_response.text()
    if response_text:
        body = json.loads(response_text)
    post_response.raise_for_status()
    if not body:
        raise ControllerApiException(f"Controller returned empty response from {url}")
    return body, False
```

### Observations

- **pg_listener retry budget**: In production, `_listen()` blocks forever on `conn.notifies()`, so after a successful retry, subsequent connection drops consume the *same* `AsyncRetrying` attempt budget rather than resetting. The test `test_backoff_resets_between_failovers` uses a finite mock iterator which masks this — each outage gets a fresh `AsyncRetrying` in the test but not in production. Low risk (10 attempts is generous), but worth noting for very long-running deployments.

- **WebSocket unbounded retries**: `_connect_websocket` retries indefinitely on transient close codes with exponential backoff (capped at 60s). No circuit breaker or retry budget. Acceptable for the current use case but could block the process indefinitely if the server persistently returns transient codes.

- **POST retry on `_launch`**: Retrying POST on 502/503 is a pragmatic trade-off. 502/503 typically mean the request didn't reach the backend, but in edge cases the gateway could return 502 after the controller accepted the launch, risking duplicates. No server-side idempotency is apparent.

### Nits (open from prior reviews)

- `tests/e2e/utils.py:167` — `connection_attempts: list = None` should be `list | None = None` (PEP 484)
- `tests/test_controller.py:816-852` — uses manual `os.environ` try/finally instead of `monkeypatch`; doesn't close the runner session
- `tests/e2e/utils.py:212` — `print(data["bad"])` to force errors is brittle; a `raise RuntimeError(...)` would be more explicit

Co-Authored-By: Claude Sonnet 4.5 <noreply@anthropic.com>

File: tests/test_controller.py [hunk 4]
```diff
@@ -783,3 +776,125 @@ async def test_get_job_url_from_label_invalid_type(new_job_template_runner):
...
+
+@pytest.mark.asyncio
+async def test_controller_unavailable_and_restored_logging(caplog):
+    """Verify unavailable/restored logging."""
+    import os
+
+    from ansible_rulebook.job_template_runner import JobTemplateRunner
+
+    old_val = os.environ.get("EDA_CONTROLLER_RETRY_ATTEMPTS")
+    os.environ["EDA_CONTROLLER_RETRY_ATTEMPTS"] = "1"
```
Comment (ptoscano) [inline tests/test_controller.py]: Simply use `monkeypatch.setenv()` for this, as done in other tests.

File: requirements_test.txt [hunk 5]
```diff
@@ -21,6 +21,7 @@ freezegun
 oauthlib>=3.2.0
 kubernetes
 urllib3 >= 2.6.3
-aioresponses
+# TODO: revert to 'aioresponses' after https://github.com/pnuckowski/aioresponses/issues/289 is released
+aioresponses @ git+https://github.com/agners/aioresponses.git@8f40e0db2eed16621607ecfc857168240c795755
```
Comment (ptoscano) [inline requirements_test.txt]: This commit doesn't exist in the actual aioresponses.git repository, but in a fork (and thus can disappear any time).

I'd rather copy https://github.com/j7an/dep-rank/pull/123 locally in our `conftest.py`, so we can still use an official version of `aioresponses`.

File: ansible_rulebook/job_template_runner.py [hunk 6]
```diff
@@ -86,16 +89,41 @@ def host(self, value: str):
...
+            self._post_retry_options = ExponentialRetry(
+                attempts=int(os.getenv("EDA_CONTROLLER_RETRY_ATTEMPTS", "5")),
+                start_timeout=2.0,
+                max_timeout=settings.controller_retry_max_timeout,
+                factor=2.0,
+                statuses={429, 502, 503},
+                retry_all_server_errors=False,
+                exceptions={aiohttp.ClientConnectorError},
+                methods={"POST"},
+            )
```
Comment (ptoscano) [inline ansible_rulebook/job_template_runner.py:130]: The only difference between these two `ExponentialRetry` objects is that the one for POST does not handle the states 500 and 504. Doesn't POST need to handle 500 and 504?

Let's use a single object instead, if there is no reason to duplicate them.

File: ansible_rulebook/job_template_runner.py [hunk 7]
```diff
@@ -86,16 +89,41 @@ def host(self, value: str):
...
         if self._session is None:
             limit = int(os.getenv("EDA_CONTROLLER_CONNECTION_LIMIT", "30"))
-            self._session = aiohttp.ClientSession(
+            client_session = aiohttp.ClientSession(
                 connector=aiohttp.TCPConnector(limit=limit),
                 headers=self._auth_headers(),
                 auth=self._basic_auth(),
+            )
+            retry_options = ExponentialRetry(
+                attempts=int(os.getenv("EDA_CONTROLLER_RETRY_ATTEMPTS", "5")),
```
Comment (ptoscano) [inline ansible_rulebook/job_template_runner.py]: Shouldn't it be better to set this as additional field in `setting`, as done with `controller_retry_max_timeout`?
Does it need to be also a new CLI option?

File: ansible_rulebook/job_template_runner.py (hunk 6 above)
Comment (mkanoor) [reply ansible_rulebook/job_template_runner.py:130]: @ptoscano When we are creating an object in the server and it issues a 500 we dont know what it has created before it crashed. Retrying on 500 with POST is risky it could relaunch multiple jobs but with GET its ok since GET is an idempotent call. Initially I had one Retry but had to change to 2 because of the POST issue.

File: ansible_rulebook/job_template_runner.py (hunk 7 above)
Comment (mkanoor) [reply ansible_rulebook/job_template_runner.py]: Added to CLI and env var

File: tests/test_controller.py [hunk 8]
```diff
@@ -783,3 +776,120 @@ async def test_get_job_url_from_label_invalid_type(new_job_template_runner):
...
+    """Verify _launch checks status before parsing JSON on 5xx."""
+    monkeypatch.setattr(settings, "controller_retry_attempts", 1)
+    monkeypatch.setattr(settings, "controller_retry_max_timeout", 0.0)
+    new_job_template_runner._create_session()
+    launch_url = "api/v2/job_templates/1/launch/"
+    with aioresponses() as mocked:
+        url = f"{new_job_template_runner.host}{launch_url}"
+        mocked.post(url, status=502, body="")
+        with pytest.raises(ControllerApiException):
+            await new_job_template_runner._launch({}, url)
```
[the remark this answers is not in the document]
Comment (mkanoor) [reply tests/test_controller.py]: Fixed

File: docs/usage.rst [hunk 9]
```diff
@@ -73,8 +75,12 @@ The `ansible-rulebook` CLI supports the following options:
                             The file containing one ansible vault password, can also be passed via the env var EDA_VAULT_PASSWORD_FILE.
     --vault-id VAULT_ID   label@filename pointing to an ansible vault password file
     --ask-vault-pass      Ask vault password interactively 
-    --persistence-id   PERSISTENCE_ID 
+    --persistence-id   PERSISTENCE_ID
                          The unique id, preferably a UUID to track persistent event data
+    --controller-retry-max-timeout CONTROLLER_RETRY_MAX_TIMEOUT
+                            Maximum backoff time in seconds for controller API retries on transient errors (429/502/503/504). Default is 60. Can also be passed via env var EDA_CONTROLLER_RETRY_MAX_TIMEOUT
+    --controller-retry-attempts CONTROLLER_RETRY_ATTEMPTS
+                            Number of retry attempts for controller API calls on transient errors. Default is 5. Can also be passed via env var EDA_CONTROLLER_RETRY_ATTEMPTS
```
[the remark this answers is not in the document]
Comment (mkanoor) [reply docs/usage.rst:83]: Fixed

File: ansible_rulebook/job_template_runner.py [hunk 10]
```diff
@@ -396,12 +455,27 @@ async def _launch(self, job_params: dict, url: str) -> dict:
...
             ) as post_response:
-                body = json.loads(await post_response.text())
+                response_text = await post_response.text()
+                if not response_text:
+                    raise ControllerApiException(
+                        f"Controller returned empty response from {url}"
+                    )
+                body = json.loads(response_text)
                 post_response.raise_for_status()
                 return body
```
[the remark this answers is not in the document]
Comment (mkanoor) [reply ansible_rulebook/job_template_runner.py:470]: Fixed

File: ansible_rulebook/event_source/pg_listener.py [hunk 11]
```diff
@@ -138,16 +191,95 @@ def _validate_args(args: dict[str, Any]) -> None:
...
+            LOGGER.warning("PostgreSQL connection lost (%s)", e)
+            async for attempt in AsyncRetrying(
+                retry=retry_if_exception(_is_retryable_pg_error),
+                wait=wait_exponential(multiplier=1, min=2, max=max_timeout),
+                stop=stop_after_attempt(max_attempts),
+                before_sleep=_log_pg_retry,
+            ):
+                with attempt:
+                    await _listen(queue, args, True)
+            reconnecting = False
```
[the remark this answers is not in the document]
Comment (mkanoor) [reply ansible_rulebook/event_source/pg_listener.py]: Fixed

**Comment by mkanoor** [comment]
@wfealdel Please review updated

File: ansible_rulebook/event_source/pg_listener.py [hunk 12]
```diff
@@ -138,36 +191,137 @@ def _validate_args(args: dict[str, Any]) -> None:
...
+                reraise=True,
+            ):
+                with attempt:
+                    conn = await _connect_and_subscribe(args, reconnecting)
+            await _process_notifications(conn, queue)
+            return
+        except asyncio.CancelledError:
+            LOGGER.info("pg_listener shutdown requested")
+            raise
+        except OperationalError as e:
```
Comment (wfealdel) [inline ansible_rulebook/event_source/pg_listener.py:249]: When a connection drops mid-chunk during `_process_notifications`, the partially accumulated `chunked_cache` is silently discarded (it's created fresh each call). Consider logging a warning here so operators can diagnose missing events after a failover, e.g.:

```suggestion
            LOGGER.warning("PostgreSQL connection lost (%s); "
                           "any in-progress chunked messages were discarded", e)
```

File: ansible_rulebook/event_source/pg_listener.py [hunk 13]
```diff
@@ -138,36 +191,137 @@ def _validate_args(args: dict[str, Any]) -> None:
...
+PG_RETRY_ATTEMPTS_DEFAULT = 10
+
+PG_KEEPALIVES_DEFAULT = 1
+PG_KEEPALIVES_IDLE_DEFAULT = 10
+PG_KEEPALIVES_INTERVAL_DEFAULT = 10
+PG_KEEPALIVES_COUNT_DEFAULT = 3
+
+
 async def main(queue: asyncio.Queue[Any], args: dict[str, Any]) -> None:
     """Listen for events from a channel."""
```
Comment (wfealdel) [inline ansible_rulebook/event_source/pg_listener.py:207]: No validation prevents an operator from setting `retry_max_timeout: 0`, which clamps `wait_exponential(min=2, max=0)` to zero-delay retries — potentially hammering a recovering database. Consider clamping to a minimum or warning:

```suggestion
    max_timeout = float(
        args.get("retry_max_timeout", PG_RETRY_MAX_TIMEOUT_DEFAULT)
    )
    if max_timeout < 2.0:
        LOGGER.warning(
            "retry_max_timeout=%.1f is below minimum backoff (2s), "
            "retries will have minimal delay",
            max_timeout,
        )
```

**Review by wfealdel — COMMENTED** [review]
@mkanoor 
Overall the retry logic across all three connection paths looks solid. Two minor suggestions on pg_listener resilience — see inline comments.

**Review by wfealdel — DISMISSED** [review]
LGTM!

**Review by ptoscano — DISMISSED** [review]
LGTM now with the changes, thanks!

File: ansible_rulebook/event_source/pg_listener.py (hunk 12 above)
Comment (mkanoor) [reply ansible_rulebook/event_source/pg_listener.py:249]: Updated

File: ansible_rulebook/event_source/pg_listener.py [hunk 14]
```diff
@@ -138,36 +191,149 @@ def _validate_args(args: dict[str, Any]) -> None:
...
+    conn = await AsyncConnection.connect(
+        conninfo=args.get("dsn", ""),
+        autocommit=True,
+        **_build_connect_params(args),
+    )
+    if reconnecting:
+        LOGGER.warning("PostgreSQL connection re-established")
+    cursor = conn.cursor()
+    for channel in args["channels"]:
+        await cursor.execute(f"LISTEN {channel};")
```
[the remark this answers is not in the document]
Comment (mkanoor) [reply ansible_rulebook/event_source/pg_listener.py]: Fixed

File: ansible_rulebook/event_source/pg_listener.py [hunk 15]
```diff
@@ -138,36 +192,152 @@ def _validate_args(args: dict[str, Any]) -> None:
...
+
+def _is_retryable_pg_error(exc: BaseException) -> bool:
+    """Retry on transient OperationalErrors but not on auth failures."""
+    return isinstance(exc, OperationalError) and not isinstance(
+        exc, (InvalidPassword, InvalidAuthorizationSpecification)
+    )
+
+
+def _log_pg_retry(retry_state) -> None:
+    exc = retry_state.outcome.exception()
```
Comment (ttuffin) [inline ansible_rulebook/event_source/pg_listener.py:271]: Since we're logging the exception here, is there a risk that we may expose some sensitive Postgres DSN info?

File: ansible_rulebook/job_template_runner.py [hunk 16]
```diff
@@ -120,9 +155,33 @@ async def _get_page(self, href_slug: str, params: dict) -> dict:
...
+            )
         except aiohttp.ClientError as e:
-            logger.error(CLIENT_CONNECT_ERROR_STRING, str(e))
+            if self._controller_available:
+                self._controller_available = False
+                logger.warning(
+                    "Controller unavailable (%s), retries exhausted",
+                    e,
+                )
+            logger.exception(CLIENT_CONNECT_ERROR_STRING, str(e))
```
Comment (ttuffin) [inline ansible_rulebook/job_template_runner.py]: Similar to the question above - since we're changing from logger.error to logger.exception, do we risk exposing sensitive connection secrets in the stack trace? Same concern with the changes below.

**Review by wfealdel — DISMISSED** [review]

**Review by wfealdel — APPROVED** [review]

File: ansible_rulebook/job_template_runner.py (hunk 16 above)
Comment (mkanoor) [reply ansible_rulebook/job_template_runner.py]: @ttuffin I changed it to 
```
logger.error(CLIENT_CONNECT_ERROR_STRING, str(e))  # NOSONAR
```

File: ansible_rulebook/event_source/pg_listener.py (hunk 15 above)
Comment (mkanoor) [reply ansible_rulebook/event_source/pg_listener.py:271]: @ttuffin I am trying to collect the logs from my testing this one should be ok since its just printing the exception as string and no stack trace

Comment (mkanoor) [reply ansible_rulebook/event_source/pg_listener.py:271]: @ttuffin 
```
<run_path> - WARNING - PostgreSQL connection lost (consuming input failed: SSL SYSCALL error: Connection timed out); any in-progress messages were discarded
<run_path> - WARNING - PostgreSQL connection re-established
```

**Review by ttuffin — APPROVED** [review]
